### PR TITLE
Use the new cached task counts in CalculateTaskPlanStats

### LINKED ISSUE #470: CalculateTaskPlanStats performance improvements
Use the new cached task counts [here](https://github.com/openstax/tutor-server/blob/d7176f0378559d5ae54b51f6844f9ac97882c959/app/routines/calculate_task_plan_stats.rb#L133-L140) and [here](https://github.com/openstax/tutor-server/blob/d7176f0378559d5ae54b51f6844f9ac97882c959/app/routines/calculate_task_plan_stats.rb#L82-L86) instead of iterating through all the steps

### PR DESCRIPTION
**Pull request by karenc**
Close #470

### CODE REVIEW

File: app/routines/calculate_task_plan_stats.rb [hunk 1]
```diff
@@ -130,13 +122,13 @@ def generate_period_stat_data
 
         total_count: period_tasks.count,
 
-        complete_count: period_tasks.count{|task|
-          task.task_steps.all?{| ts | ts.completed? }
+        complete_count: period_tasks.count { |task|
```
Comment (jpslav) [inline app/routines/calculate_task_plan_stats.rb]: There's already a `complete?` method which does this comparison so you can just say:

```ruby
complete_count: period_tasks.count(&:complete?)
```

File: app/routines/calculate_task_plan_stats.rb [hunk 2]
```diff
@@ -130,13 +122,13 @@ def generate_period_stat_data
...
+          task.completed_steps_count == task.steps_count
         },
 
-        partially_complete_count: period_tasks.count{|task|
-          task.task_steps.any?{| ts | ts.completed? } &&
-            !task.task_steps.all?{| ts | ts.completed? }
+        partially_complete_count: period_tasks.count { |task|
+          task.completed_steps_count != 0 &&
+            task.completed_steps_count != task.steps_count
```
Comment (jpslav) [inline app/routines/calculate_task_plan_stats.rb]: Add an `in_progress?` method to `Tasks::Models::Task` which has this logic and then make this be:

```ruby
partially_complete_count: period_tasks.count(&:in_progress?)
```

**Comment by jpslav** [comment]
Review complete

**Comment by karenc** [comment]
Ready for review again.

**Comment by Dantemss** [comment]
I think I have an idea to speed this up some more... I'll post some comments.

**Comment by Dantemss** [comment]
In https://github.com/openstax/tutor-server/blob/stats-improvements/app/routines/calculate_task_plan_stats.rb#L110 make it `.eager_load([task_steps, {taskings: :period}])`

In https://github.com/openstax/tutor-server/blob/stats-improvements/app/routines/calculate_task_plan_stats.rb#L88 change it to something similar to https://github.com/openstax/tutor-server/blob/master/lib/course_guide_methods.rb#L8 (probably don't need all of that eager_load, only the exercise -> page stuff)

In https://github.com/openstax/tutor-server/blob/stats-improvements/app/routines/calculate_task_plan_stats.rb#L92 use the page preloaded in line 88 above.

In https://github.com/openstax/tutor-server/blob/master/app/subsystems/tasks/models/task.rb#L113 remove the calls to preload the `:tasked` (might need to add it elsewhere, not too sure, but not for the stats)

**Comment by Dantemss** [comment]
Hopefully this should get rid of the N+1 queries...

**Comment by karenc** [comment]
Thanks!  I'll start working on it.

**Comment by jpslav** [comment]
@karenc are you working on Dante's suggestion in this branch or elsewhere?

**Comment by karenc** [comment]
@jpslav sorry I got distracted, I plan to work on Dante's suggestion in this branch.

**Comment by jpslav** [comment]
@karenc going to do @Dantemss's extra optimizations in a later PR.

**Pull request merged**